### PR TITLE
Running a single node on k8s

### PR DESCRIPTION
**Pull request by krish7919**
*  with distinct mongodb and bigchaindb pods on k8s
*  service discovery using kube-dns between pods
*  changing the directory structure and adding the yaml files to repo
*  using separate storage classes to map 2 different volumes exposed by the mongo docker container; one for /data/db (dbPath) and the other for /data/configdb (configDB).
*  using separate pvc as azure allows using a claim only once per container in a node.
*  corresponding documentation

### CODE REVIEW

File: docs/server/source/cloud-deployment-templates/node-on-kubernetes.rst [hunk 1]
```diff
@@ -159,36 +164,102 @@ but it should become "Bound" fairly quickly.
...
 
-   $ wget https://raw.githubusercontent.com/bigchaindb/bigchaindb/master/k8s/node-mdb-ss.yaml
+   $ wget https://raw.githubusercontent.com/bigchaindb/bigchaindb/master/k8s/mongodb/mongo-ss.yaml
+
+
+Note how the MongoDB container uses the ``mongo-db-claim`` and the
+``mongo-configdb-claim`` PersistentVolumeClaims for its ``/data/db`` and
+``/data/configdb`` diretories (mount path).
+
+Note also that we use the pod's ``securityContext.capabilities.add`` specification
```
Comment (ttmc) [inline docs/server/source/cloud-deployment-templates/node-on-kubernetes.rst]: This sentence and the next few are about the same thing so should be combined into one paragraph: just remove the blank space between them.

File: docs/server/source/cloud-deployment-templates/node-on-kubernetes.rst [hunk 2]
```diff
@@ -159,36 +164,102 @@ but it should become "Bound" fairly quickly.
...
+Refer to the
+`Docker doc <https://docs.docker.com/engine/reference/run/#runtime-privilege-and-linux-capabilities>`_
+for details.
 
-Take a look inside that file to see how it defines the Service
-and the StatefulSet.
-Note how the MongoDB container uses the ``mongoclaim`` PersistentVolumeClaim
-for its ``/data`` diretory (mount path).
+As we gain more experience running MongoDB in testing and production, we will
+tweak the ``resources.limits`` values for cpu and memory.
```
Comment (ttmc) [inline docs/server/source/cloud-deployment-templates/node-on-kubernetes.rst]: cpu --> CPU here and elsewhere.

File: docs/server/source/cloud-deployment-templates/node-on-kubernetes.rst [hunk 3]
```diff
@@ -159,36 +164,102 @@ but it should become "Bound" fairly quickly.
...
 .. code:: bash
 
-   $ kubectl get services
-   $ kubectl get statefulsets
+   $ kubectl apply -f bigchaindb-dep.yaml
+
+You can check its status using the command ``kubectl get deploy -w``
+
+
+Step 7: Verifying our node is setup correctly
```
Comment (ttmc) [inline docs/server/source/cloud-deployment-templates/node-on-kubernetes.rst]: Titles should be written in title case, and should take an imperative "Do This" form, e.g. "Step 7: Verify the BigchainDB Node Setup"

File: docs/server/source/cloud-deployment-templates/template-kubernetes-azure.rst [hunk 4]
```diff
@@ -113,8 +115,26 @@ go to **Resource groups** (with the blue cube icon)
 and click on the one you created
 to see all the resources in it.
 
-Next, you can :doc:`run a BigchainDB node on your new
-Kubernetes cluster <node-on-kubernetes>`.
+
+Step 3: (Optional) Delete the managed k8s cluster
```
Comment (ttmc) [inline docs/server/source/cloud-deployment-templates/template-kubernetes-azure.rst]: Please write all titles in title case.

File: docs/server/source/cloud-deployment-templates/template-kubernetes-azure.rst [hunk 5]
```diff
@@ -141,5 +160,31 @@ the master (a bad idea),
...
+   $ echo -e "Host <FQDN of the cluster from Azure Portal>\n  ForwardAgent yes" >> ~/.ssh/config
+
+To verify whether SSH Forwarding works properly, login to the one of the master
+machines and run
+
+.. code:: bash
+
+   $ echo "$SSH_AUTH_SOCK"
+
+If you get an empty response, SSH forwarding hasn't been set up correctlt.
```
Comment (ttmc) [inline docs/server/source/cloud-deployment-templates/template-kubernetes-azure.rst]: correctlt --> correctly

File: docs/server/source/cloud-deployment-templates/template-kubernetes-azure.rst (hunk 4 above)
Comment (ttmc) [inline docs/server/source/cloud-deployment-templates/template-kubernetes-azure.rst]: I'd rename this step as just "Optional: Delete the Kubernetes Cluster"

File: docs/server/source/cloud-deployment-templates/template-kubernetes-azure.rst [hunk 6]
```diff
@@ -113,8 +115,26 @@ go to **Resource groups** (with the blue cube icon)
...
+-------------------------------------------------
+
+.. code:: bash
+
+   $ az acs delete \
+   --name <ACS cluster name> \
+   --resource-group <name of resource group containing the cluster> 
+
+
+Step 4: (Optional) Delete the resource group
```
Comment (ttmc) [inline docs/server/source/cloud-deployment-templates/template-kubernetes-azure.rst]: I'd rename this step as just "Optional: Delete the Resource Group"

**Comment by ttmc** [comment]
The two optional sections, about how to delete the k8s cluster or delete the resource group, should be moved to the end of the page, since you won't delete them and then SSH to them.

File: docs/server/source/cloud-deployment-templates/node-on-kubernetes.rst [hunk 7]
```diff
@@ -67,25 +75,30 @@ the PersistentVolumeClaim would get stuck in a "Pending" state.
...
-     name: slow
-   provisioner: kubernetes.io/azure-disk
-   parameters:
-     skuName: Standard_LRS
-     location: <region where your cluster is located>
+.. code:: bash
 
-and then:
+   $ wget https://raw.githubusercontent.com/bigchaindb/bigchaindb/master/k8s/mongodb/mongo-data-db-sc.yaml
+   $ wget https://raw.githubusercontent.com/bigchaindb/bigchaindb/master/k8s/mongodb/mongo-data-configdb-sc.yaml
```
Comment (krish7919) [inline docs/server/source/cloud-deployment-templates/node-on-kubernetes.rst]: Can we do that as a part of the next PR. I had thought about it, but it makes sense to do it step by step at the first iteration and then make it shorter?

File: docs/server/source/cloud-deployment-templates/node-on-kubernetes.rst [hunk 8]
```diff
@@ -128,21 +133,21 @@ read-write by a single Kubernetes node.
 by AzureDisk.)
 ``storage: 20Gi`` means the volume has a size of 20
 `gibibytes <https://en.wikipedia.org/wiki/Gibibyte>`_.
-(You can change that if you like.)
 
-Create ``mongoclaim`` in your Kubernetes cluster:
+You may want to update the ``spec.resources.requests.storage`` field in both
+the files to specify a different disk size.
```
Comment (krish7919) [inline docs/server/source/cloud-deployment-templates/node-on-kubernetes.rst]: Well, I had referred this link: https://kubernetes.io/docs/user-guide/persistent-volumes/

> Currently, only NFS and HostPath support recycling. AWS EBS, GCE PD, Azure Disk, and Cinder volumes support deletion.

File: docs/server/source/cloud-deployment-templates/node-on-kubernetes.rst (hunk 2 above)
Comment (krish7919) [inline docs/server/source/cloud-deployment-templates/node-on-kubernetes.rst]: Well, it is `resources.limits.cpu` in the config. Anyway, I will modify it.

File: docs/server/source/cloud-deployment-templates/node-on-kubernetes.rst (hunk 7 above)
Comment (ttmc) [inline docs/server/source/cloud-deployment-templates/node-on-kubernetes.rst]: ok

File: docs/server/source/cloud-deployment-templates/node-on-kubernetes.rst (hunk 8 above)
Comment (ttmc) [inline docs/server/source/cloud-deployment-templates/node-on-kubernetes.rst]: That doesn't explain very well what the different reclaim policies actually mean.

File: k8s/mongodb/mongo-data-configdb-pvc.yaml [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+kind: PersistentVolumeClaim
+apiVersion: v1
+metadata:
+  name: mongo-configdb-claim
+  annotations:
+    volume.beta.kubernetes.io/storage-class: slow-configdb
+spec:
+  accessModes:
+  - ReadWriteOnce
+  persistentVolumeReclaimPolicy: Retain
```
Comment (krish7919) [inline k8s/mongodb/mongo-data-configdb-pvc.yaml]: From https://kubernetes.io/docs/user-guide/persistent-volumes/.
> Currently, only NFS and HostPath support recycling. AWS EBS, GCE PD, Azure Disk, and Cinder volumes support deletion.

I can specify it, but the UI still says that the reclaim policy is `Deletion`

File: docs/server/source/cloud-deployment-templates/node-on-kubernetes.rst (hunk 2 above)
Comment (ttmc) [inline docs/server/source/cloud-deployment-templates/node-on-kubernetes.rst]: Oh, maybe just write:

```text
tweak the ``resource.limits.cpu`` and ``resources.limits.memory``.
```

File: k8s/mongodb/mongo-data-configdb-pvc.yaml (hunk 9 above)
Comment (ttmc) [inline k8s/mongodb/mongo-data-configdb-pvc.yaml]: The quoted sentence only says that Azure Disk supports deletion. It doesn't say if Azure Disk supports `Retain`. (It says nothing about support or lack of support.)

Anyway, if `Retain` works on Azure, then keep it in: that's what we want.

**Comment by ttmc** [comment]
Do we want to merge any of the files in the `bigchaindb/k8s/deprecated.to.del` directory?

**Comment by krish7919** [comment]
It will be useful to refer to for some time initially. Once we get
comfortable, we can delete the dir.
It has lots of interesting snippets - working ones.
Will delete it in a separate commit.




Best,
Krish

File: docs/server/source/cloud-deployment-templates/node-on-kubernetes.rst [hunk 10]
```diff
@@ -128,21 +133,21 @@ read-write by a single Kubernetes node.
...
+   $ kubectl apply -f mongo-data-db-pvc.yaml
+   $ kubectl apply -f mongo-data-configdb-pvc.yaml
 
-.. code:: bash
 
-   $ kubectl get pvc
+You can check its status using: ``kubectl get pvc -w``
 
-Initially, the status of ``mongoclaim`` might be "Pending"
+Initially, the status of persistent volume claims might be "Pending"
```
Comment (ttmc) [inline docs/server/source/cloud-deployment-templates/node-on-kubernetes.rst]: Please remove or update the block of text below this (i.e. showing `$ kubectl describe pvc` and its output): it is out-of-date.

**Comment by ttmc** [comment]
This subsection title is getting broken into two:

docs1

It can probably be shortened to just "Optional: Set Up SSH Forwarding"

File: docs/server/source/cloud-deployment-templates/template-kubernetes-azure.rst [hunk 11]
```diff
@@ -141,5 +139,52 @@ the master (a bad idea),
...
+
+CAUTION: You might end up deleting resources other than the ACS cluster.
+
+.. code:: bash
+
+   $ az group delete \
+   --name <name of resource group containing the cluster>
+
+
+Next, you can :doc:`run a BigchainDB node on your new
```
Comment (ttmc) [inline docs/server/source/cloud-deployment-templates/template-kubernetes-azure.rst]: For some reason, this sentence at the end of the page is repeated twice.

**Pull request merged**